File: src/main/clients/openrgb/effects/impl/SpectrumCycle.ts
import Device from '../../client/classes/Device';
import { RGBColor } from '../../client/classes/RGBColor';
import Client from '../../client/client';
import { AbstractEffect } from '../AbstractEffect';

export class SpectrumCycle extends AbstractEffect {
  public getName(): string {
    return 'Spectrum Cycle';
  }

  protected async applyEffect(client: Client, devices: Array<Device>): Promise<void> {
    for (let offset = 0; this.isRunning; offset = (offset + 1) % 360) {
      devices.forEach((element, i) => {
        client.updateLeds(
          i,
          Array(element.leds.length).fill(RGBColor.fromHSV(offset, 1, this.brightness))
        );
      });
      await new Promise<void>((resolve) => setTimeout(resolve, 40));
    }
    this.hasFinished = true;
  }
}

File: rcc/communications/client/tcp/openrgb/effects/spectrum_cycle.py
from openrgb.utils import RGBColor

from framework.singleton import singleton
from rcc.communications.client.tcp.openrgb.effects.base.abstract_effect import AbstractEffect


@singleton
class SpectrumCycle(AbstractEffect):
    """Spectrum Cycle efect"""

    def __init__(self):
        super().__init__("Spectrum cycle")

    def apply_effect(self):
        offset = 0
        while self._is_running:
            for dev in self._devices:
                if dev.enabled:
                    colors = [RGBColor.fromHSV(offset, 100, 100)] * len(dev.leds)
                    self._set_colors(dev, colors)
            offset = (offset + 1) % 360
            self._sleep(0.02)  # 20 ms


SPECTRUM_CYCLE_EFFECT = SpectrumCycle()
